pkg/vsphere/tasks/waiter.go

```diff
@@ -278,6 +278,41 @@ func IsNotFoundError(err error) bool {
 	return false
 }
 
+// IsInvalidPowerStateError is an error certifier function for errors coming back from vsphere. It checks for an InvalidPowerStateFault
+func IsInvalidPowerStateError(err error) bool {
+	if soap.IsVimFault(err) {
+		_, ok1 := soap.ToVimFault(err).(*types.InvalidPowerState)
+		_, ok2 := soap.ToVimFault(err).(*types.InvalidPowerStateFault)
+		return ok1 || ok2
+	}
+
+	if soap.IsSoapFault(err) {
+		_, ok1 := soap.ToSoapFault(err).VimFault().(types.InvalidPowerState)
+		_, ok2 := soap.ToSoapFault(err).VimFault().(types.InvalidPowerStateFault)
+		// sometimes we get the correct fault but wrong type
+		return ok1 || ok2 || soap.ToSoapFault(err).String == "vim.fault.InvalidPowerState" ||
+			soap.ToSoapFault(err).String == "vim.fault.InvalidPowerState"
+	}
+	return false
+}
+
+// IsInvalidStateError is an error certifier function for errors coming back from vsphere. It checks for an InvalidStateFault
+func IsInvalidStateError(err error) bool {
+	if soap.IsVimFault(err) {
+		_, ok1 := soap.ToVimFault(err).(*types.InvalidState)
```

Comment:
In the other similar methods, we also check for the `*types.____Fault` here. Is there a reason we don't check for `types.InvalidStateFault` in this case? (If so, that would be good to document in a comment here so that future readers don't assume it's a mistake.)

Reply:
iirc `types.InvalidStateFault` does not implement `GetMethodFault()` method so it won't compile. I confess I've not dug into why this one is not the same as the others. @dougm ?